### PR TITLE
KAFKA-3888 Use background thread to process consumer heartbeats

### PR DESCRIPTION
**Pull request by dpkp**
This PR implements KAFKA-3888 (including related KAFKA-4431). It adds significant complexity and relies on managing shared mutable state, which I really don't feel great about. Nonetheless, I wanted to put this up so that folks can see where it's at and possibly test out independently.

One remaining issue: address default configuration changes when broker does not support rebalance timeout (only session timeout). 

ref #948 / includes #1258 
See also KAFKA-4160

### CODE REVIEW

File: test/test_coordinator.py [hunk 1]
```diff
@@ -551,28 +560,18 @@ def test_send_offset_fetch_request_success(patched_coord, partitions):
         future, response) 
 
 
-@pytest.mark.parametrize('response,error,dead,reassign', [
-    #(OffsetFetchResponse[0]([('foobar', [(0, 123, b'', 30), (1, 234, b'', 30)])]),
-    # Errors.GroupAuthorizationFailedError, False, False),
-    #(OffsetFetchResponse[0]([('foobar', [(0, 123, b'', 7), (1, 234, b'', 7)])]),
-    # Errors.RequestTimedOutError, True, False),
-    #(OffsetFetchResponse[0]([('foobar', [(0, 123, b'', 27), (1, 234, b'', 27)])]),
-    # Errors.RebalanceInProgressError, False, True),
```
Comment (jeffwidman) [inline test/test_coordinator.py]: What was the initial purpose of these? Why were they commented out?

**Comment by tvoinarovskyi** [comment]
Will try to do a review tomorrow. Sorry for the holdup.

File: kafka/client_async.py [hunk 2]
```diff
@@ -512,46 +527,40 @@ def poll(self, timeout_ms=None, future=None, delayed_tasks=True):
         Returns:
             list: responses received (can be empty)
         """
-        if timeout_ms is None:
+        if future is not None:
+            timeout_ms = 100
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/client_async.py]: Yes, deserves a comment. The issue is that we now have to account for the future being resolved in a different thread. Prior to this change we would block on network IO and then check the future's completion after processing the network IO and any corresponding futures. But now that we have separated future handling from network IO, and we have (intentionally) moved the future handling into a separate section that is not locked, we can have a situation where one thread is called with a simple timeout and another is called with a future (block until resolved). The first thread acquires the lock and receives the response that will resolve the future in its _poll() call. The response is put on the internal pending_completions queue and the lock is released. Now the second thread takes the lock and checks whether the future is resolved. It is not, so it calls _poll with a full request timeout, which is now 5 minutes. The first thread continues, without the lock, and begins processing pending_completions. The future is resolved with the queued response, and the first thread finishes. But the second thread is now waiting for network IO that it thinks is necessary to complete its future. Except that the future is already done and so there is no network IO coming. It will simply wait for 5 minutes and then timeout before rechecking the future and then finally returning.

My solution here is to simply reduce the network IO timeout when called with a future to reduce the unneeded block time when this occurs. Another solution might be to try to register any future that is "blocking" and call _wake_up() on when such a future is resolved. I decided against that approach because I think it is too complex and it is likely to add overhead to every response, not just ones for which some thread is blocking on a future.

File: kafka/client_async.py [hunk 3]
```diff
@@ -721,6 +739,7 @@ def _maybe_refresh_metadata(self):
         Returns:
             int: milliseconds until next refresh
         """
+        # This should be locked when running multi-threaded
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/client_async.py]: Comment relates to entire function, which is currently only called inside a locked section. Will update.

File: kafka/client_async.py [hunk 4]
```diff
@@ -772,34 +791,6 @@ def refresh_done(val_or_error):
...
-        """Unschedule a task.
-
-        This will remove all instances of the task from the task queue.
-        This is a no-op if the task is not scheduled.
-
-        Arguments:
-            task (callable): task to be unscheduled
-        """
-        self._delayed_tasks.remove(task)
-
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/client_async.py]: The other "tasks" are metadata refresh and offset commits. These are now handled inline during client.poll and coordinator.poll, respectively.

File: kafka/coordinator/base.py [hunk 5]
```diff
@@ -242,52 +282,116 @@ def need_rejoin(self):
         """
         return self.rejoin_needed
 
+    def poll_heartbeat(self):
+        """
+        Check the status of the heartbeat thread (if it is active) and indicate
+        the liveness of the client. This must be called periodically after
+        joining with :meth:`.ensureActiveGroup` to ensure that the member stays
+        in the group. If an interval of time longer than the provided rebalance
+        timeout expires without calling this method, then the client will
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: Good call. It is actually called rebalance_timeout at the protocol level, but the consumer configuration that drives it is max_poll_interval_ms.

File: kafka/coordinator/base.py [hunk 6]
```diff
@@ -242,52 +282,116 @@ def need_rejoin(self):
...
+            self.join_future = None
+            self.state = MemberState.STABLE
+            self.rejoining = False
+            self._heartbeat_thread.enable()
+        self._on_join_complete(self._generation.generation_id,
+                               self._generation.member_id,
+                               self._generation.protocol,
+                               member_assignment_bytes)
+
+    def _handle_join_failure(self, exception):
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: Yea, it's just because that is the errback interface. I'll change it to `_` to make it clear that we're ignoring it intentionally.

File: kafka/coordinator/base.py [hunk 7]
```diff
@@ -309,14 +413,35 @@ def _send_join_group_request(self):
...
-            self.group_id,
-            self.config['session_timeout_ms'],
-            self.member_id,
-            self.protocol_type(),
-            [(protocol,
-              metadata if isinstance(metadata, bytes) else metadata.encode())
-             for protocol, metadata in self.group_protocols()])
+        member_metadata = [
+            (protocol, metadata if isinstance(metadata, bytes) else metadata.encode())
+            for protocol, metadata in self.group_protocols()
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: I think this is described in the docstring of `group_protocols`

[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: For further information, see kafka/coordinator/assignors/* -- we support assignors returning a struct that can be encoded, like `ConsumerProtocolMemberMetadata` or just raw bytes.

File: kafka/coordinator/base.py [hunk 8]
```diff
@@ -758,6 +843,112 @@ def __init__(self, heartbeat, metrics, prefix, tags=None):
...
+                # poll again after waiting for the retry backoff in case
+                # the heartbeat failed or the coordinator disconnected
+                log.debug('Not ready to heartbeat, waiting')
+                self.coordinator._lock.wait(self.coordinator.config['retry_backoff_ms'] / 1000)
+
+            else:
+                self.coordinator.heartbeat.sent_heartbeat()
+                future = self.coordinator._send_heartbeat_request()
+                future.add_callback(self._handle_heartbeat_success)
+                future.add_errback(self._handle_heartbeat_failure)
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: Taken straight from java client. I think I changed 'heartbeat.send_heartbeat()` to `heartbeat.sent_heartbeat()` just to keep myself somewhat sane. But the flow and internal logic are the same.

File: kafka/coordinator/consumer.py [hunk 9]
```diff
@@ -104,13 +107,14 @@ def __init__(self, client, subscription, metrics, **configs):
...
+                self.next_auto_commit_deadline = time.time() + self.auto_commit_interval
 
         self.consumer_sensors = ConsumerCoordinatorMetrics(
             metrics, self.config['metric_group_prefix'], self._subscription)
 
+        self._cluster.request_update()
+        self._cluster.add_listener(WeakMethod(self._handle_metadata_update))
+
     def __del__(self):
         if hasattr(self, '_cluster') and self._cluster:
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/consumer.py]: Yes, this could be `is not None`. But we probably should move away from using `__del__` directly and consider `weakref.finalize` instead

File: kafka/coordinator/consumer.py [hunk 10]
```diff
@@ -227,6 +230,54 @@ def _on_join_complete(self, generation, member_id, protocol,
...
+            # initial rebalance, we need to ensure that the metadata is fresh
+            # before joining initially, and then request the metadata update. If
+            # metadata update arrives while the rebalance is still pending (for
+            # example, when the join group is still inflight), then we will lose
+            # track of the fact that we need to rebalance again to reflect the
+            # change to the topic subscription. Without ensuring that the
+            # metadata is fresh, any metadata update that changes the topic
+            # subscriptions and arrives with a rebalance in progress will
+            # essentially be ignored. See KAFKA-3949 for the complete
+            # description of the problem.
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/consumer.py]: No, that will need a separate PR

File: kafka/coordinator/consumer.py [hunk 11]
```diff
@@ -227,6 +230,54 @@ def _on_join_complete(self, generation, member_id, protocol,
...
+            self.ensure_active_group()
+
+        self.poll_heartbeat()
+        self._maybe_auto_commit_offsets_async()
+
+    def time_to_next_poll(self):
+        """
+        Return the time to the next needed invocation of {@link #poll(long)}.
+        @param now current time in milliseconds
+        @return the maximum time in milliseconds the caller should wait before the next invocation of poll()
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/consumer.py]: +1

File: kafka/coordinator/consumer.py [hunk 12]
```diff
@@ -344,6 +395,14 @@ def close(self, autocommit=True):
         finally:
             super(ConsumerCoordinator, self).close()
 
+    def _invoke_completed_offset_commit_callbacks(self):
+        try:
+            while True:
+                callback, offsets, exception = self.completed_offset_commits.popleft()
+                callback(offsets, exception)
+        except IndexError:
+            pass
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/consumer.py]: yes, that seems cleaner

File: kafka/errors.py [hunk 13]
```diff
@@ -59,7 +59,17 @@ class UnrecognizedBrokerVersion(KafkaError):
...
+    def __init__(self, *args, **kwargs):
+        super(CommitFailedError, self).__init__(
+            """Commit cannot be completed since the group has already
+            rebalanced and assigned the partitions to another member.
+            This means that the time between subsequent calls to poll()
+            was longer than the configured max.poll.interval.ms, which
+            typically implies that the poll loop is spending too much
+            time message processing. You can address this either by
+            increasing the session timeout or by reducing the maximum
+            size of batches returned in poll() with max.poll.records.
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/errors.py]: Good call on updating the text. But the CommitFailed error is only raised when using group coordinated groups It will not be seen when using zookeeper offsets from 0.8.2. That said, it can be seen by users that do not have max_poll_interval_ms support (0.9 <= broker < 0.10.1), and so it may need some tweaking when we decide how to manage max_poll_interval_ms in that context.

File: test/test_coordinator.py (hunk 1 above)
Comment (dpkp) [reply test/test_coordinator.py]: I'm not sure and so I deleted them.

File: kafka/coordinator/base.py [hunk 14]
```diff
@@ -758,6 +843,112 @@ def __init__(self, heartbeat, metrics, prefix, tags=None):
...
+
+            if self.coordinator.state is not MemberState.STABLE:
+                # the group is not stable (perhaps because we left the
+                # group or because the coordinator kicked us out), so
+                # disable heartbeats and wait for the main thread to rejoin.
+                log.debug('Group state is not stable, disabling heartbeats')
+                self.disable()
+                return
+
+            # When consumer.wakeup() is implemented, we need to
```
Comment (tvoinarovskyi) [inline kafka/coordinator/base.py]: Add XXX or FIXME here, so we don't lose it.

File: kafka/coordinator/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tvoinarovskyi) [reply kafka/coordinator/base.py]: Yea, you can get that from reading Java code. It never shrinks, parts are only added on top. Probably because a lot of people are working on it, but really is hard to follow sometimes.

**Comment by tvoinarovskyi** [comment]
Great job on this!

File: kafka/client_async.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffwidman) [reply kafka/client_async.py]: I am glad I asked. I never would have realized that.

[the remark this answers is not in the document]
Comment (tvoinarovskyi) [reply kafka/client_async.py]: Same =)

File: kafka/coordinator/base.py [hunk 15]
```diff
@@ -242,52 +282,116 @@ def need_rejoin(self):
...
+            # we do not call it a second time if the client is woken up before
+            # a pending rebalance completes.
+            if not self.rejoining:
+                self._on_join_prepare(self._generation.generation_id,
+                                      self._generation.member_id)
+                self.rejoining = True
+
+            if self._heartbeat_thread is None:
+                log.debug('Starting new heartbeat thread')
+                self._heartbeat_thread = HeartbeatThread(weakref.proxy(self))
```
[the remark this answers is not in the document]
Comment (dpkp) [reply kafka/coordinator/base.py]: There is a circular reference between coordinator <-> heartbeat_thread. The weakref here means that the existence of the heartbeat_thread will not prevent the coordinator from being garbage collected.

**Comment by dpkp** [comment]
So this is a very large PR. I think it would be possible to break it into a few smaller parts (locks, MemberState/Generation, coordinator.poll, heartbeat_thread). The benefit of keeping it together is that it is easier to cross-reference with the java PR that implements the same.

Re testing, I think the existing test suite covers a fair amount of the raw functionality. But we should add some specific tests for heartbeating while not polling. Separately, I am very interested in refactoring to a sans-io core that can be more easily unit tested without requiring fixtures or mocking. But I'll leave that for another day (I've been working on a simple state-machine representation of the group coordinator in a separate branch).

**Comment by tvoinarovskyi** [comment]
@dpkp - is the configuration change backward compatible? If I have a consumer configured with a custom 'session_timeout_ms' will it not break after update? As far as I see it will require to add 'max_poll_ms'...

**Comment by tvoinarovskyi** [comment]
As for the merge, I think it's ok as a big PR. It is not something that has meaning in merging separately.

**Comment by dpkp** [comment]
@tvoinarovskyi - I put the compatibility logic into KafkaConsumer and left the Coordinator classes to simply enforce constraints. KafkaConsumer will check the api_version after KafkaClient bootstrap, and if < 0.10.1 (max_poll_interval_ms not supported in JoinGroupRequest) then it will set max_poll_interval_ms = session_timeout_ms . In addition, if the user provides max_poll_interval_ms but not session_timeout_ms, we use the same value for session_timeout_ms. If the user provides neither, we use 30000/30000 defaults instead of the 10000/300000 used when max_poll_interval_ms is supported.

**Comment by jeffwidman** [comment]
What are next steps here? @dpkp are there still features/tests you're planning to add when you get time or is this ready to merge to master?

**Comment by dpkp** [comment]
I had hoped to deploy this build to production in some isolated services and verify that it performs as expected. I haven't had a chance to do that yet. Other than that I think it's ready to land.

**Comment by jeffwidman** [comment]
Sounds good. I've been hoping to do a few impromptu tests myself, particularly in a couple of services we have that use gevent, but haven't had a chance either...

**Comment by tvoinarovskyi** [comment]
@dpkp Hey there, this PR will only get harder to rebase later. Is there any reason besides testing to not merge it? If it's merged we can at least find some feedback from others if it works as expected.

**Comment by dpkp** [comment]
I think only because I have been travelling and offline quite a bit, so
would not be able to provide quick support if bugs we're found. I am ok
with merging.

**Comment by dpkp** [comment]
I rebased + fixed conflicts. I'm planning to merge after travis tests pass

**Comment by jeffwidman** [comment]
Thanks again for all the hard work on this massive PR.

**Comment by aswinjoseroy** [comment]
Sorry but I am unable to understand if this is available on PyPi, kafka-python version 1.4.1?

**Comment by sanfilippopablo** [comment]
1. Is this already out in the wild in some kafka-python version?
2. From what I understand, this will allow me to have long processing times without kafka removing the consumer from the consumer group without having to set the session timeout to the worst case processing time. Is that right?

**Comment by tvoinarovskyi** [comment]
@sanfilippopablo 
1) Yes, as of 1.4.0
2) Yes, your assumption is correct

**Comment by vkjv** [comment]
May be its quite long, but I am running into the same issue with 1.4.4 client. I am using iterator interface to receive messages, which is in a separate thread. And the processing is done in another thread. All the configuration parameters set to default except for request_timeout which is set to 60seconds. 
Every 5 minutes, I get the error heartbeat expired and consumer is dead. 
And messages have been lost. 
How do I get over it?

**Comment by jeffwidman** [comment]
@vkjv this is a closed issue, you are better off opening a new one.
